vllm/worker/model_runner.py

```diff
@@ -365,20 +365,45 @@ def _compute_for_prefix_cache_hit(
                             and self.sliding_window is None
                             and inter_data.is_prompt)
         inter_data.prefix_cache_hit = prefix_cache_hit
-        if self.chunked_prefill_enabled and prefix_cache_hit:
+        if (self.chunked_prefill_enabled and prefix_cache_hit
+                and self.attn_backend.get_name() != "flash-attn"):
             raise RuntimeError(
-                "chunked prefill cannot be used with prefix caching now.")
+                "Chunked prefill and prefix caching can only be used "
+                "simultaneously with flash-attn backend, try switching "
+                "to flash-attn backend by setting the environment variable "
+                "\"VLLM_ATTENTION_BACKEND=FLASH_ATTN\"")
 
         # If prefix cache is hit, advance context length to bypass
         # hit blocks. Accordingly, input tokens, position and query length
         # have to be updated.
         if prefix_cache_hit:
             assert computed_block_nums is not None
-            context_len = len(computed_block_nums) * self.block_size
-            inter_data.input_tokens[seq_idx] = inter_data.input_tokens[
-                seq_idx][context_len:]
-            inter_data.input_positions[seq_idx] = inter_data.input_positions[
-                seq_idx][context_len:]
+            prefix_cache_len = len(computed_block_nums) * self.block_size
+            # When prefix caching meets chunked prefill, we would be in
+            # the following three conditions:
+            #
+            #   - prefix_cache_len <= context_len:
+            #     - do normal chunked prefill and nothing special
+            #   - context_len < prefix_cache_len < seq_len:
+            #     - advance the context_len to seq_len to perform non-
+            #       cached parts of the sequence.
+            #   - prefix_cache_len >= seq_len:
+            #     - it means the current partial sequence is fully cache
+            #       hit, and no further computation is needed.
```

Comment:
nit
```suggestion
            # When prefix caching meets chunked prefill, we would be in
            # one of the following three cases:
            #   - prefix_cache_len <= context_len:
            #     Do normal chunked prefill
            #   - context_len < prefix_cache_len < seq_len:
            #     Advance the context_len to seq_len to prefill non-
            #     cached parts of the sequence.
            #   - prefix_cache_len >= seq_len:
            #     The current partial sequence is fully cache hit,
            #     and no further computation is needed.
```

Another way is embedding the comments to each branch:

```python
if context_len < prefix_cache_len < seq_len:
    # prefix_cache_len <= context_len: Do normal chunked prefill
    ....
elif seq_len <= prefix_cache_len:
    # context_len < prefix_cache_len < seq_len: ...
    ...
else:
    # prefix_cache_len >= seq_len:: ...
    pass
```

I'm fine with either way